oci-image/src/pull.rs

```diff
@@ -0,0 +1,188 @@
+use crate::state::LayerMetadata;
+use crate::{Error, MetadataManager, Result, State};
+use flate2::read::GzDecoder;
+use oci_distribution::manifest::OciManifest;
+use oci_distribution::secrets::RegistryAuth;
+use oci_distribution::{Client, Reference};
+use sha2::Digest;
+use std::fs::create_dir_all;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tar::Archive;
+use tokio::sync::Mutex;
+
+/// `Puller` is responsible about pull OCI images from registries.
+///
+/// It provides methods to simply pull an image on your machine.
+pub(crate) struct Puller {
+    /// The OCI client which will be used to interact with registries.
+    client: Client,
+
+    /// The OCI registry auth information.
+    auth: RegistryAuth,
+
+    /// The OCI image reference
+    reference: Reference,
+
+    /// The directory where the image will be stored
+    image_dir: PathBuf,
+
+    /// The accepted media types which will be required to pull images.
+    /// Example :
+    accepted_media_types: Vec<String>,
+}
+
+impl Puller {
+    /// Create a new instance of `Puller`.
+    ///
+    /// Configure client, authentication and image reference
+    /// in order to properly pull the image.
+    pub fn new(image: &str, image_dir: &Path) -> Result<Self> {
+        let reference = Reference::try_from(image)
+            .map_err(|_| Error::InvalidOCIReference(image.to_string()))?;
+        let auth = RegistryAuth::Anonymous;
+
+        Ok(Self {
+            auth,
+            reference,
+            client: Client::default(),
+            image_dir: image_dir.to_path_buf(),
+            accepted_media_types: vec![],
+        })
+    }
+
+    /// Pull the image manifest and the configuration from the registry.
+    pub async fn pull_manifest(&mut self) -> Result<(OciManifest, String, String)> {
+        let (manifest, digest, config) = self
+            .client
+            .pull_manifest_and_config(&self.reference, &self.auth)
+            .await
+            .map_err(|e| Error::PullManifest(e.to_string()))?;
+
+        // Build a list of the accepted media types
+        self.accepted_media_types = manifest
+            .layers
+            .iter()
+            .map(|x| x.media_type.clone())
+            .collect::<Vec<String>>();
+
+        // Remove duplicates
+        self.accepted_media_types.dedup();
+
+        Ok((manifest, digest, config))
+    }
+
+    /// Pull the image from the registry. The method will return the complete path to the image on disk.
```

Comment:
I dont think this comment is accurate. This method returns a vector of layer metadata, which is a good thing.